drivers/samsung/displays/mdc_protocol.cr

```diff
@@ -0,0 +1,376 @@
+        end
+      when .speaker?
+        self[:speaker] = SpeakerModes.from_value(value).to_s
+      when .hard_off?
+        self[:hard_off] = value == 0
+        self[:power] = false if self[:hard_off]
+      when .screen_split?
+        self[:screen_split] = value >= 0
+      when .software_version?
+        self[:software_version] = values.join
+      when .serial_number?
+        self[:serial_number] = values.join
+      else
+        logger.debug { "Samsung responded with ACK: #{value}" }
+      end
+
+      task.try &.success
+    when .nak?
+      task.try &.abort("Samsung responded with NAK: #{hex}")
+    else
+      task.try &.abort("Samsung aborted with: #{hex}")
+    end
+  end
+
+  def check_power_state
+    return if @power_stable
+    if self[:power]? == self[:power_target]?
+      @power_stable = true
+    else
+      power(self[:power_target].as_bool)
+    end
+  end
+
+  enum Command
+    Status           = 0x00
+    Hard_Off         = 0x11 # Completely powers off
+    Panel_Mute       = 0xF9 # Screen blanking / visual mute
+    Volume           = 0x12
+    Contrast         = 0x24
+    Brightness       = 0x25
+    Sharpness        = 0x26
+    Colour           = 0x27
+    Tint             = 0x28
+    Red_Gain         = 0x29
+    Green_Gain       = 0x2A
+    Blue_Gain        = 0x2B
+    Input            = 0x14
+    Mode             = 0x18
+    Size             = 0x19
+    Pip              = 0x3C # picture in picture
+    Auto_Adjust      = 0x3D
+    Wall_Mode        = 0x5C # Video wall mode
+    Safety           = 0x5D
+    Wall_On          = 0x84 # Video wall enabled
+    Wall_User        = 0x89 # Video wall user control
+    Speaker          = 0x68
+    Network_Standby  = 0xB5 # Keep NIC active in standby, enable power on (without WOL)
+    Auto_Off_Timer   = 0xE6 # Eco options (auto power off)
+    Auto_Power       = 0x33 # Device auto power control (presumably signal based?)
+    Screen_Split     = 0xB2 # Tri / quad split (larger panels only)
+    Software_Version = 0x0E
+    Serial_Number    = 0x0B
+    Time             = 0xA7
+    Timer            = 0xA4
+
+    def build(id : Int32, data : Bytes) : Bytes
+      Bytes.new(data.size + 5).tap do |bytes|
+        bytes[0] = INDICATOR                            # Header
+        bytes[1] = self.to_u8                           # Command
+        bytes[2] = id.to_u8                             # Display ID
+        bytes[3] = data.size.to_u8                      # Data size
+        data.each_with_index(4) { |b, i| bytes[i] = b } # Data
+        bytes[-1] = bytes[1..-2].reduce(&.+)            # Checksum
+      end
+    end
+  end
+
+  private def do_send(command : Command, data : Int | Bytes = Bytes.empty, **options)
+    data = Bytes[data] if data.is_a?(Int)
+    bytes = command.build(@id, data)
+    logger.debug { "Sending to Samsung: #{bytes.hexstring}" }
+    send(bytes, **options)
+    # TODO: find out if this is necessary
```

Comment:
TODO needs to be resolved before merge.

Reply:
removed as it's from the old ruby implementation